From 85a624e20d5419eeb134641ba517207b1b285cac Mon Sep 17 00:00:00 2001
From: Arien Judge <arien.judge@morsemicro.com>
Date: Fri, 12 Jul 2024 17:22:20 +1000
Subject: [PATCH] board: morsemicro: add MorseMicro HaLowLink 1 board

This adds the basic set of board files for HaLowLink 1, based on the MT7621

Signed-off-by: Arien Judge <arien.judge@morsemicro.com>
---
 arch/mips/dts/Makefile               |   1 +
 arch/mips/dts/morsemicro,halowlink1.dts  | 163 +++++++++++++++++++++++++++
 arch/mips/mach-mtmips/mt7621/Kconfig |  12 ++
 board/morsemicro/halowlink1/Kconfig      |  21 ++++
 board/morsemicro/halowlink1/Makefile     |   3 +
 board/morsemicro/halowlink1/board.c      |   1 +
 board/morsemicro/halowlink1/env          |  23 ++++
 configs/morsemicro_halowlink1_defconfig  | 100 ++++++++++++++++
 8 files changed, 324 insertions(+)
 create mode 100644 arch/mips/dts/morsemicro,halowlink1.dts
 create mode 100644 board/morsemicro/halowlink1/Kconfig
 create mode 100644 board/morsemicro/halowlink1/Makefile
 create mode 100644 board/morsemicro/halowlink1/board.c
 create mode 100644 board/morsemicro/halowlink1/env
 create mode 100644 configs/morsemicro_halowlink1_defconfig

Index: u-boot-2024.10/arch/mips/dts/Makefile
===================================================================
--- u-boot-2024.10.orig/arch/mips/dts/Makefile
+++ u-boot-2024.10/arch/mips/dts/Makefile
@@ -15,6 +15,7 @@ dtb-$(CONFIG_BOARD_COMTREND_VR3032U) +=
 dtb-$(CONFIG_BOARD_COMTREND_WAP5813N) += comtrend,wap-5813n.dtb
 dtb-$(CONFIG_BOARD_HUAWEI_HG556A) += huawei,hg556a.dtb
 dtb-$(CONFIG_BOARD_INTENO_XG6846) += inteno,xg6846.dtb
+dtb-$(CONFIG_BOARD_MORSEMICRO_HALOWLINK1) += morsemicro,halowlink1.dtb
 dtb-$(CONFIG_BOARD_MT7620_RFB) += mediatek,mt7620-rfb.dtb
 dtb-$(CONFIG_BOARD_MT7620_MT7530_RFB) += mediatek,mt7620-mt7530-rfb.dtb
 dtb-$(CONFIG_BOARD_MT7621_RFB) += mediatek,mt7621-rfb.dtb
Index: u-boot-2024.10/arch/mips/dts/morsemicro,halowlink1.dts
===================================================================
--- /dev/null
+++ u-boot-2024.10/arch/mips/dts/morsemicro,halowlink1.dts
@@ -0,0 +1,163 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright (C) 2024 Morse Micro
+ *
+ * Author: Arien Judge <arien.judge@morsemicro.com>
+ */
+
+/dts-v1/;
+
+#include "mt7621.dtsi"
+#include <dt-bindings/gpio/gpio.h>
+
+/ {
+	compatible = "morse,halowlink1", "mediatek,mt7621-soc";
+	model = "MorseMicro HalowLink 1";
+
+	aliases {
+		serial0 = &uart0;
+		spi0 = &spi;
+	};
+
+	chosen {
+		stdout-path = &uart0;
+	};
+
+	keys {
+		compatible = "gpio-keys";
+
+		reset {
+			label = "reset";
+			gpios = <&gpio0 18 GPIO_ACTIVE_LOW>;
+		};
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		led0_red: led0_red {
+			label = "red:led0";
+			gpios = <&gpio0 13 GPIO_ACTIVE_LOW>;
+			default-state = "on";
+		};
+
+		led0_green: led0_green {
+			label = "green:led0";
+			gpios = <&gpio0 14 GPIO_ACTIVE_LOW>;
+			default-state = "off";
+		};
+
+		led0_blue: led0_blue {
+			label = "blue:led0";
+			gpios = <&gpio0 15 GPIO_ACTIVE_LOW>;
+			default-state = "off";
+		};
+
+		led1_red: led1_red {
+			label = "red:led1";
+			gpios = <&gpio0 23 GPIO_ACTIVE_LOW>;
+			default-state = "off";
+		};
+
+		led1_green: led1_green {
+			label = "green:led1";
+			gpios = <&gpio0 22 GPIO_ACTIVE_LOW>;
+			default-state = "off";
+		};
+
+		led1_blue: led1_blue {
+			label = "blue:led1";
+			gpios = <&gpio1 1 GPIO_ACTIVE_LOW>;
+			default-state = "on";
+		};
+
+		led2_red: led2_red {
+			label = "red:led2";
+			gpios = <&gpio0 28 GPIO_ACTIVE_LOW>;
+			default-state = "off";
+		};
+
+		led2_green: led2_green {
+			label = "green:led2";
+			gpios = <&gpio0 29 GPIO_ACTIVE_LOW>;
+			default-state = "on";
+		};
+
+		led2_blue: led2_blue {
+			label = "blue:led2";
+			gpios = <&gpio1 0 GPIO_ACTIVE_LOW>;
+			default-state = "off";
+		};
+	};
+};
+
+&pinctrl {
+	state_default: pin_state {
+		groups = "wdt", "uart3", "jtag", "uart2", "rgmii2";
+		function = "gpio";
+	};
+
+	eth_pins: eth_pins {
+		mdio_pins {
+				groups = "mdio";
+				function = "mdio";
+			};
+		esw_pins {
+			groups = "esw int";
+			function = "esw int";
+		};
+	};
+};
+
+&uart0 {
+	status = "okay";
+};
+
+&gpio {
+	status = "okay";
+};
+
+&spi {
+	status = "okay";
+	num-cs = <2>;
+
+	spi-flash@0 {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		compatible = "jedec,spi-nor";
+		spi-max-frequency = <50000000>;
+		reg = <0>;
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				label = "u-boot";
+				reg = <0x0 0x80000>;
+			};
+
+			partition@80000 {
+				label = "u-boot-env";
+				reg = <0x80000 0x10000>;
+			};
+
+			partition@90000 {
+				label = "factory";
+				reg = <0x90000 0x10000>;
+			};
+
+			partition@A0000 {
+				label = "firmware";
+				reg = <0xA0000 0x1f60000>;
+			};
+		};
+	};
+};
+
+&eth {
+	/* Just enable reset and mdio pins */
+	pinctrl-names = "default";
+	pinctrl-0 = <&eth_pins>;
+	status = "okay";
+};
Index: u-boot-2024.10/arch/mips/mach-mtmips/mt7621/Kconfig
===================================================================
--- u-boot-2024.10.orig/arch/mips/mach-mtmips/mt7621/Kconfig
+++ u-boot-2024.10/arch/mips/mach-mtmips/mt7621/Kconfig
@@ -99,6 +99,16 @@ config BOARD_MT7621_NAND_RFB
 	  sockets, 1 RGMII to external GbE PHY, 2 audio jacks (in/out),
 	  JTAG pins and expansion GPIO pins.
 
+config BOARD_MORSEMICRO_HALOWLINK1
+	bool "Morse Micro HaLowLink 1"
+	help
+	  Morse Micro HaLowLink 1 boards have an MT7621A booting from 32MB SPI-NOR flash.
+	  The boards have 128MB DDR3 RAM.
+	  Between its different peripherals there is an AzureWave AW-HM593 providing
+	  HaLow Wi-Fi connectivity, an RTL8152B providing a connected host with
+	  ethernet over USB. and an MT7603 providing 802.11n Wi-Fi connectivity.
+	  Other peripherals include - 3 RGB LEDs, 2 ethernet ports, and a push button.
+
 endchoice
 
 config SYS_CONFIG_NAME
@@ -112,3 +122,5 @@ config SYS_VENDOR
 	default "mediatek" if BOARD_MT7621_RFB || BOARD_MT7621_NAND_RFB
 
 endif
+
+source "board/morsemicro/halowlink1/Kconfig"
Index: u-boot-2024.10/board/morsemicro/halowlink1/Kconfig
===================================================================
--- /dev/null
+++ u-boot-2024.10/board/morsemicro/halowlink1/Kconfig
@@ -0,0 +1,21 @@
+if BOARD_MORSEMICRO_HALOWLINK1
+
+config SYS_BOARD
+	default "halowlink1"
+
+config SYS_VENDOR
+	default "morsemicro"
+
+config SYS_CONFIG_NAME
+	default "mt7621"
+
+config DEFAULT_DEVICE_TREE
+	default "morsemicro,halowlink1"
+
+config USE_DEFAULT_ENV_FILE
+	default y
+
+config DEFAULT_ENV_FILE
+	default "morse_halowlink1_env"
+
+endif
Index: u-boot-2024.10/board/morsemicro/halowlink1/Makefile
===================================================================
--- /dev/null
+++ u-boot-2024.10/board/morsemicro/halowlink1/Makefile
@@ -0,0 +1,3 @@
+# SPDX-License-Identifier: GPL-2.0
+
+obj-y += board.o
Index: u-boot-2024.10/board/morsemicro/halowlink1/board.c
===================================================================
--- /dev/null
+++ u-boot-2024.10/board/morsemicro/halowlink1/board.c
@@ -0,0 +1 @@
+// SPDX-License-Identifier: GPL-2.0
Index: u-boot-2024.10/board/morsemicro/halowlink1/env
===================================================================
--- /dev/null
+++ u-boot-2024.10/board/morsemicro/halowlink1/env
@@ -0,0 +1,23 @@
+ipaddr=192.168.12.1
+serverip=192.168.12.2
+loadaddr=0x83000000
+flashaddr=0xbfca0000
+bootcmd=run boot_firmware
+bootconf=config-1
+bootdelay=3
+bootled_red=red:led0
+bootled_green=green:led0
+bootmenu_confirm_return=askenv - Press ENTER to return to menu ; bootmenu 60
+bootmenu_default=0
+bootmenu_delay=3
+bootmenu_title=HaLowLink 1 U-Boot
+bootmenu_0=Run default boot command.=run bootcmd
+bootmenu_1=Boot system via TFTP.=run boot_tftp
+bootmenu_2=Load system via TFTP then write to flash.=run flash_tftp
+bootmenu_3=Reboot.=reset
+bootmenu_4=Reset uboot settings to factory defaults.=run reset_factory ; reset
+boot_tftp=askenv ipaddr Device IP address:; askenv serverip Server IP Address:; askenv bootfile Filename on server:; tftpboot ${loadaddr} ${bootfile} && bootm ${loadaddr}
+boot_firmware=bootm ${flashaddr}
+flash_tftp=askenv ipaddr Device IP address:; askenv serverip Server IP Address:; askenv bootfile Filename on server:; tftpboot ${loadaddr} ${bootfile} && run nor_write_firmware
+reset_factory=mtd erase u-boot-env
+nor_write_firmware=mtd erase firmware && mtd write firmware ${loadaddr}
\ No newline at end of file
Index: u-boot-2024.10/configs/morsemicro_halowlink1_defconfig
===================================================================
--- /dev/null
+++ u-boot-2024.10/configs/morsemicro_halowlink1_defconfig
@@ -0,0 +1,100 @@
+CONFIG_MIPS=y
+CONFIG_SYS_HAS_NONCACHED_MEMORY=y
+CONFIG_SYS_MALLOC_LEN=0x100000
+CONFIG_SPL_LIBCOMMON_SUPPORT=y
+CONFIG_SPL_LIBGENERIC_SUPPORT=y
+CONFIG_NR_DRAM_BANKS=1
+CONFIG_SF_DEFAULT_SPEED=20000000
+CONFIG_ENV_SIZE=0x1000
+CONFIG_SPL_SERIAL=y
+CONFIG_DM_SERIAL=y
+CONFIG_SERIAL_RX_BUFFER=y
+CONFIG_SERIAL_RX_BUFFER_SIZE=256
+CONFIG_TPL_SYS_MALLOC_F_LEN=0x1000
+CONFIG_SPL_SYS_MALLOC_F_LEN=0x40000
+CONFIG_SPL=y
+CONFIG_DEBUG_UART_BASE=0xbe000c00
+CONFIG_DEBUG_UART_CLOCK=50000000
+CONFIG_SYS_LOAD_ADDR=0x83000000
+CONFIG_ARCH_MTMIPS=y
+CONFIG_SOC_MT7621=y
+CONFIG_BOARD_MORSEMICRO_HALOWLINK1=y
+CONFIG_DEFAULT_ENV_FILE="board/morsemicro/halowlink1/env"
+CONFIG_SYS_MIPS_TIMER_FREQ=440000000
+# CONFIG_MIPS_CACHE_SETUP is not set
+# CONFIG_MIPS_CACHE_DISABLE is not set
+CONFIG_RESTORE_EXCEPTION_VECTOR_BASE=y
+CONFIG_MIPS_BOOT_FDT=y
+CONFIG_DEBUG_UART=y
+CONFIG_FIT=y
+# CONFIG_ARCH_FIXUP_FDT_MEMORY is not set
+CONFIG_SPI_BOOT=y
+CONFIG_AUTOBOOT_KEYED=y
+CONFIG_AUTOBOOT_MENU_SHOW=y
+CONFIG_SYS_CONSOLE_INFO_QUIET=y
+CONFIG_SPL_MAX_SIZE=0x30000
+CONFIG_SPL_BSS_START_ADDR=0x80140000
+CONFIG_SPL_BSS_MAX_SIZE=0x80000
+CONFIG_SPL_SYS_MALLOC_SIMPLE=y
+CONFIG_SPL_NOR_SUPPORT=y
+CONFIG_SPL_SPI_FLASH_MTD=y
+CONFIG_TPL=y
+# CONFIG_TPL_FRAMEWORK is not set
+# CONFIG_BOOTM_NETBSD is not set
+# CONFIG_BOOTM_PLAN9 is not set
+# CONFIG_BOOTM_RTEMS is not set
+# CONFIG_BOOTM_VXWORKS is not set
+CONFIG_SYS_BOOTM_LEN=0x2000000
+CONFIG_CMD_BOOTMENU=y
+# CONFIG_CMD_ELF is not set
+# CONFIG_CMD_XIMG is not set
+CONFIG_CMD_ASKENV=y
+# CONFIG_CMD_CRC32 is not set
+CONFIG_CMD_GPIO=y
+# CONFIG_CMD_LOADS is not set
+CONFIG_CMD_MTD=y
+CONFIG_CMD_PART=y
+# CONFIG_CMD_PINMUX is not set
+CONFIG_CMD_SPI=y
+CONFIG_CMD_DHCP=y
+CONFIG_CMD_TFTPSRV=y
+# CONFIG_CMD_NFS is not set
+CONFIG_CMD_DNS=y
+CONFIG_CMD_LINK_LOCAL=y
+CONFIG_DOS_PARTITION=y
+# CONFIG_SPL_DOS_PARTITION is not set
+# CONFIG_ISO_PARTITION is not set
+CONFIG_EFI_PARTITION=y
+# CONFIG_SPL_EFI_PARTITION is not set
+CONFIG_PARTITION_TYPE_GUID=y
+CONFIG_SYS_RELOC_GD_ENV_ADDR=y
+CONFIG_NET_RANDOM_ETHADDR=y
+CONFIG_BUTTON=y
+CONFIG_BUTTON_GPIO=y
+CONFIG_GPIO_HOG=y
+# CONFIG_I2C is not set
+# CONFIG_INPUT is not set
+CONFIG_LED=y
+CONFIG_LED_BLINK=y
+CONFIG_LED_GPIO=y
+CONFIG_MTD=y
+CONFIG_SPI_FLASH_BAR=y
+CONFIG_SPI_FLASH_EON=y
+CONFIG_SPI_FLASH_GIGADEVICE=y
+CONFIG_SPI_FLASH_ISSI=y
+CONFIG_SPI_FLASH_MACRONIX=y
+CONFIG_SPI_FLASH_SPANSION=y
+CONFIG_SPI_FLASH_STMICRO=y
+CONFIG_SPI_FLASH_WINBOND=y
+CONFIG_SPI_FLASH_XMC=y
+CONFIG_SPI_FLASH_XTX=y
+CONFIG_SPI_FLASH_MTD=y
+CONFIG_MEDIATEK_ETH=y
+CONFIG_PHY=y
+CONFIG_PHY_MTK_TPHY=y
+CONFIG_DEBUG_UART_SHIFT=2
+CONFIG_SPI=y
+CONFIG_MT7621_SPI=y
+CONFIG_SYSRESET=y
+CONFIG_SYSRESET_RESETCTL=y
+# CONFIG_BINMAN_FDT is not set
+CONFIG_LZMA=y
+# CONFIG_GZIP is not set
+CONFIG_SPL_LZMA=y
